File: examples/fragment_service/fragment_service_with_resource/python/fragment_service_with_resource.py
"""
SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
SPDX-License-Identifier: Apache-2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

import weakref

from holoscan.conditions import CountCondition
from holoscan.core import (
    Application,
    DefaultFragmentService,
    FragmentService,
    Operator,
    OperatorSpec,
    Resource,
)


class MyServiceResource(Resource):
    """A resource that can be used as a fragment service."""

    def __init__(self, fragment, *args, int_value=0, **kwargs):
        self.int_value = int_value

        super().__init__(fragment, *args, **kwargs)

    def setup(self, spec):
        # In Python, we don't use spec.param for simple attributes
        # Instead, we just store the value directly
        pass

    def value(self):
        """Get the integer value stored in the resource."""
        return self.int_value


class MyResourceManagerSelfRef(Resource, FragmentService):
    """A resource that also implements FragmentService interface with self-reference."""

    def __init__(self, fragment, *args, int_value=0, **kwargs):
        self.int_value = int_value

        Resource.__init__(self, fragment, *args, **kwargs)
        FragmentService.__init__(self)
        self._resource_ref = weakref.ref(self)

    def resource(self, new_resource=None):
        """Overrides the C++ resource property."""
        if new_resource is not None:
            # This is the setter path, called from C++
            self._resource_ref = weakref.ref(new_resource)
            # Also call the C++ base class's resource setter to keep it in sync
            super().resource(new_resource)
        # The getter path
        return self._resource_ref()

    def value(self):
        """Get the integer value stored in the resource."""
        return self.int_value


class MyResourceManagerEnabledShared(Resource, FragmentService):
    """A resource that also implements FragmentService interface."""

    def __init__(self, fragment, *args, int_value=0, **kwargs):
        self.int_value = int_value
        Resource.__init__(self, fragment, *args, **kwargs)
        FragmentService.__init__(self)

    def resource(self, _new_resource=None):
        """Overrides the C++ resource property. The setter is a no-op."""
        # The setter is called from C++ during registration, but we don't need to
        # store the new_resource since this object is its own resource.
        return self

    def value(self):
        """Get the integer value stored in the resource."""
        return self.int_value


class MyOp(Operator):
    """An operator that retrieves and uses fragment services."""

    def setup(self, spec: OperatorSpec):
        pass

    def compute(self, op_input, op_output, context):
        print("MyOp.compute() executed")

        # 1,2) Fragment service resource can be retrieved via the base DefaultFragmentService type
        my_service = self.service(DefaultFragmentService, "my_service")
        if my_service:
            resource = my_service.resource
            if hasattr(resource, "value"):
                print(f"MyService value (via FragmentService): {resource.value()}")
            else:
                print("ERROR: Resource doesn't have value() method")
        else:
            print("ERROR: Could not retrieve service via FragmentService")

        # 1,2) Fragment service resource can also be retrieved directly via the resource type
        # Note: When retrieving a resource by its specific type, use the resource name as ID
        my_service_resource = self.service(MyServiceResource, "my_service")
        if my_service_resource:
            print(f"MyService value (via MyServiceResource): {my_service_resource.value()}")
        else:
            print("ERROR: Could not retrieve service via MyServiceResource")

        # 3) Fragment service resource can be retrieved via the resource type inheriting both
        #    Resource and FragmentService interfaces (self-reference)
        my_service_resource_selfref = self.service(
            MyResourceManagerSelfRef, "my_service_resource_selfref"
        )
        if my_service_resource_selfref:
            print(
                f"MyResourceManager value (via MyResourceManagerSelfRef): "
                f"{my_service_resource_selfref.value()}"
            )
        else:
            print("ERROR: Could not retrieve service via MyResourceManagerSelfRef")

        # 4) Fragment service resource can be retrieved via the resource type inheriting both
        #    Resource and FragmentService interfaces
        my_service_resource_enabled_shared = self.service(
            MyResourceManagerEnabledShared, "my_resource_manager_enabled_shared"
        )
        if my_service_resource_enabled_shared:
            print(
                f"MyResourceManager value (via MyResourceManagerEnabledShared): "
                f"{my_service_resource_enabled_shared.value()}"
            )
        else:
            print("ERROR: Could not retrieve service via MyResourceManagerEnabledShared")


class FragmentServiceApp(Application):
    """Application demonstrating fragment service usage with resources."""

    def compose(self):
        # WARNING: The Fragment Service API is currently experimental in Holoscan SDK.
        # Future SDK releases may introduce breaking changes to this API.

        # 1,2) Create a service resource with an integer value of 20
        my_service_resource = MyServiceResource(self, name="my_service", int_value=20)

        # # 1) Fragment service can be registered with a FragmentService instance (created with
        # #    a Resource object)
        # my_service = DefaultFragmentService(my_service_resource)
        # self.register_service(my_service)  # ID automatically handled for resources

        # 2) Fragment service can also be registered directly with a Resource object
        # Note: For resources, the ID parameter is ignored and the resource's name is used
        self.register_service(my_service_resource)  # id parameter defaults to "" and can be omitted

        # 3) Create and register a fragment service that inherits from both Resource and
        #    FragmentService.
        my_service_resource_selfref = MyResourceManagerSelfRef(
            self, name="my_service_resource_selfref", int_value=30
        )
        self.register_service(my_service_resource_selfref)

        # 4) Create and register a fragment service that inherits from both Resource and
        #    FragmentService.
        my_resource_manager_enabled_shared = MyResourceManagerEnabledShared(
            self, name="my_resource_manager_enabled_shared", int_value=40
        )
        self.register_service(my_resource_manager_enabled_shared)

        # Create an operator with a CountCondition to run once
        my_op = MyOp(self, CountCondition(self, 1), name="my_op")

        # Add the operator to the application
        self.add_operator(my_op)


def main():
    app = FragmentServiceApp()
    app.run()


if __name__ == "__main__":
    main()

File: examples/fragment_service/fragment_service_with_resource/cpp/fragment_service_with_resource.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <memory>
#include <utility>

#include <holoscan/holoscan.hpp>

class MyServiceResource : public holoscan::Resource {
 public:
  HOLOSCAN_RESOURCE_FORWARD_ARGS_SUPER(MyServiceResource, holoscan::Resource)

  MyServiceResource() = default;

  void setup(holoscan::ComponentSpec& spec) override {
    spec.param(int_param_, "int_value", "Integer parameter", "Integer parameter for testing", {});
  }
  [[nodiscard]] int value() const { return int_param_.get(); }

 private:
  holoscan::Parameter<int> int_param_;
};

class MyResourceManagerSelfRef : public holoscan::Resource, public holoscan::FragmentService {
 public:
  HOLOSCAN_RESOURCE_FORWARD_ARGS_SUPER(MyResourceManagerSelfRef, holoscan::Resource)

  MyResourceManagerSelfRef() = default;

  [[nodiscard]] std::shared_ptr<Resource> resource() const override { return resource_.lock(); }

  void resource(const std::shared_ptr<Resource>& resource) override { resource_ = resource; }

  void setup(holoscan::ComponentSpec& spec) override {
    spec.param(int_param_, "int_value", "Integer parameter", "Integer parameter for testing", {});
  }

  [[nodiscard]] int value() const { return int_param_.get(); }

 private:
  holoscan::Parameter<int> int_param_;
  std::weak_ptr<Resource> resource_;  ///< Weak reference to the managed resource (self)
};

// NOLINTNEXTLINE(fuchsia-multiple-inheritance)
class MyResourceManagerEnabledShared
    : public holoscan::Resource,
      public holoscan::FragmentService,
      public std::enable_shared_from_this<MyResourceManagerEnabledShared> {
 public:
  HOLOSCAN_RESOURCE_FORWARD_ARGS_SUPER(MyResourceManagerEnabledShared, holoscan::Resource)

  MyResourceManagerEnabledShared() = default;
  ~MyResourceManagerEnabledShared() override = default;
  MyResourceManagerEnabledShared(const MyResourceManagerEnabledShared&) = delete;
  MyResourceManagerEnabledShared& operator=(const MyResourceManagerEnabledShared&) = delete;
  MyResourceManagerEnabledShared(MyResourceManagerEnabledShared&&) = delete;
  MyResourceManagerEnabledShared& operator=(MyResourceManagerEnabledShared&&) = delete;

  [[nodiscard]] std::shared_ptr<Resource> resource() const override {
    return std::const_pointer_cast<MyResourceManagerEnabledShared>(shared_from_this());
  }

  void resource([[maybe_unused]] const std::shared_ptr<Resource>& resource) override {}

  void setup(holoscan::ComponentSpec& spec) override {
    spec.param(int_param_, "int_value", "Integer parameter", "Integer parameter for testing", {});
  }

  [[nodiscard]] int value() const { return int_param_.get(); }

 private:
  holoscan::Parameter<int> int_param_;
};

class MyOp : public holoscan::Operator {
 public:
  HOLOSCAN_OPERATOR_FORWARD_ARGS(MyOp)

  MyOp() = default;

  void setup(holoscan::OperatorSpec& spec) override {}

  void compute([[maybe_unused]] holoscan::InputContext& op_input,
               [[maybe_unused]] holoscan::OutputContext& op_output,
               [[maybe_unused]] holoscan::ExecutionContext& context) override {
    HOLOSCAN_LOG_INFO("MyOp::compute() executed");

    // 1,2) Fragment service resource can be retrieved via the base FragmentService type
    auto my_service = service<holoscan::DefaultFragmentService>("my_service");
    HOLOSCAN_LOG_INFO("MyService value (via FragmentService): {}",
                      my_service->resource<MyServiceResource>()->value());

    // 1,2) Fragment service resource can also be retrieved directly via the resource type
    auto my_service_resource = service<MyServiceResource>("my_service");
    HOLOSCAN_LOG_INFO("MyService value (via Resource): {}", my_service_resource->value());

    // 3) Fragment service resource can be retrieved via the resource type inheriting both
    //    holoscan::Resource and holoscan::FragmentService interfaces (self-reference)
    auto my_service_resource_selfref =
        service<MyResourceManagerSelfRef>("my_service_resource_selfref");
    HOLOSCAN_LOG_INFO("MyResourceManager value (via MyResourceManagerSelfRef): {}",
                      my_service_resource_selfref->value());

    // 4) Fragment service resource can be retrieved via the resource type inheriting both
    //    holoscan::Resource and holoscan::FragmentService interfaces (shared_from_this())
    auto my_service_resource_enabled_shared =
        service<MyResourceManagerEnabledShared>("my_resource_manager_enabled_shared");
    HOLOSCAN_LOG_INFO("MyResourceManager value (via MyResourceManagerEnabledShared): {}",
                      my_service_resource_enabled_shared->value());
  }
};

class FragmentServiceApp : public holoscan::Application {
 public:
  void compose() override {
    using namespace holoscan;
    // WARNING: The Fragment Service API is currently experimental in Holoscan SDK.
    // Future SDK releases may introduce breaking changes to this API.

    // 1,2) Create a service resource with an integer value of 20
    auto my_service_resource =
        make_resource<MyServiceResource>("my_service", Arg("int_value") = 20);

    // // 1) Fragment service can be registered with a DefaultFragmentService instance (created with
    // //    a holoscan::Resource object)
    // auto my_service = std::make_shared<DefaultFragmentService>(my_service_resource);
    // register_service(my_service);

    // 2) Fragment service can also be registered directly with a Holoscan Resource object
    register_service(my_service_resource);

    // 3) Create and register a fragment service that inherits from both Resource and
    //    FragmentService.
    //    The service maintains a weak reference to itself that is initialized during
    //    registration.
    auto my_service_resource_selfref = make_resource<MyResourceManagerSelfRef>(
        "my_service_resource_selfref", Arg("int_value") = 30);
    register_service(my_service_resource_selfref);

    // 4) Create and register a fragment service that inherits from both Resource and
    //    FragmentService.
    //    The service maintains a shared reference to itself that is initialized during
    //    registration.
    auto my_resource_manager_enabled_shared = make_resource<MyResourceManagerEnabledShared>(
        "my_resource_manager_enabled_shared", Arg("int_value") = 40);
    register_service(my_resource_manager_enabled_shared);

    auto my_op = make_operator<MyOp>("my_op", make_condition<holoscan::CountCondition>(1));
    add_operator(my_op);
  }
};

int main([[maybe_unused]] int argc, [[maybe_unused]] char** argv) {
  auto app = holoscan::make_application<FragmentServiceApp>();
  app->run();

  return 0;
}
